plugins/modules/ndo_l3out_interface_group_policy.py

```diff
@@ -0,0 +1,630 @@
+        state=dict(type="str", default="query", choices=["absent", "query", "present"]),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "absent", ["name"]],
+            ["state", "present", ["name"]],
+        ],
+        required_one_of=[
+            ["template", "template_id"],
+            ["l3out", "l3out_uuid"],
+        ],
+        mutually_exclusive=[
+            ["template", "template_id"],
+            ["l3out", "l3out_uuid"],
+            ["interface_routing_policy", "interface_routing_policy_uuid"],
+            ["custom_qos_policy", "custom_qos_policy_uuid"],
+        ],
+    )
+
+    mso = MSOModule(module)
+
+    template = module.params.get("template")
+    template_id = module.params.get("template_id")
+    l3out = module.params.get("l3out")
+    l3out_uuid = module.params.get("l3out_uuid")
+    name = module.params.get("name")
+    description = module.params.get("description")
+    interface_routing_policy = module.params.get("interface_routing_policy")
+    interface_routing_policy_uuid = module.params.get("interface_routing_policy_uuid")
+    bfd = module.params.get("bfd")
+    bfd_multi_hop = module.params.get("bfd_multi_hop")
+    ospf = module.params.get("ospf")
+    qos_priority = module.params.get("qos_priority")
+    custom_qos_policy = module.params.get("custom_qos_policy")
+    custom_qos_policy_uuid = module.params.get("custom_qos_policy_uuid")
+    state = module.params.get("state")
+
+    mso_template = MSOTemplate(mso, "l3out", template, template_id)
+    mso_template.validate_template("l3out")
+    object_description = "L3Out Interface Group Policy"
+
+    l3out_object = mso_template.get_l3out_object(l3out_uuid, l3out, True)
+    existing_interface_groups = l3out_object.details.get("interfaceGroups", [])
+    mso_template.check_template_when_name_is_provided(mso, interface_routing_policy)
+    mso_template.check_template_when_name_is_provided(mso, custom_qos_policy)
+    reference_dict = {
+        "qos": {
+            "name": "qosName",
+            "reference": "qosRef",
+            "type": "qos",
+            "template": "qosTemplateName",
+            "templateId": "qosTemplateId",
+        },
+        "interfaceRoutingPolicy": {
+            "name": "interfaceRoutingPolicyName",
+            "reference": "interfaceRoutingPolicyRef",
+            "type": "l3OutIntfPolGroup",
+            "template": "interfaceRoutingPolicyTemplateName",
+            "templateId": "interfaceRoutingPolicyTemplateId",
+        },
+    }
+    if name:
+        match = mso_template.get_object_by_key_value_pairs(
+            object_description,
+            existing_interface_groups,
+            [KVPair("name", name)],
+        )
+        if match:
+            mso_template.set_template_id_and_name(mso_template.set_names_for_references(mso, match.details, reference_dict))
```

Comment:
I wonder if we could merge these two functions into another utility function that takes in an optional `reference_dict` to simplify our module code?